scheduler/preemption.go

```diff
@@ -435,6 +439,99 @@ OUTER:
 	return filteredBestAllocs
 }
 
+// allocDeviceGroup represents a group of allocs that share a device
+type allocDeviceGroup struct {
+	allocs []*structs.Allocation
+
+	// deviceInstances tracks the number of instances used per alloc
+	deviceInstances map[string]int
+}
+
+func newAllocDeviceGroup() *allocDeviceGroup {
+	return &allocDeviceGroup{
+		deviceInstances: make(map[string]int),
+	}
+}
+
+// PreemptForDevice tries to find allocations to preempt to meet devices needed
+// This is called once per task when assigning devices to the task
+func (p *Preemptor) PreemptForDevice(ask *structs.RequestedDevice, devAlloc *deviceAllocator) []*structs.Allocation {
+
+	// Group allocations by device, tracking the number of
+	// instances used in each device by alloc id
```

Comment:
I see I think I was thrown off by `deviceInstances` field name -- if grouping by allocid, it might make more sense to have alloc in the name - maybe `allocReservations`, `allocDeviceCount`?  Up to you.